### PR TITLE
WIP Enable users to derive extra params from params + site data

### PR DESCRIPTION
**Pull request by holyjak**
so that they can f.ex. display a list of tags with their counts.

**NOTE: This is not a finished solution** - it is intended as a base for discussion before I commit too much time [possibly going in the wrong direction] to it. So what do you think, @lacarmen ? 

## ALTERNATIVES

Instead of adding `:extend-params-fn` to the overrides, we could pass it
as an extra param but it is complicated because we need to do it on 2
functions and would then perhaps need to support 3 arities. This seems
simpler.

## TODO / CONSIDER

* Rename `overrides` to something better (what???)
* Use Spec to document the arguments to the extend-params-fn?
* Document extend-params-fn in docstrings and perhaps the site docs?

### CODE REVIEW

**Comment by holyjak** [comment]
(BTW, I am migrating my blog from GatsbyJS, which puts all data into GraphQL (which has some strong downsides). This proposal 👆is a minimal change but if we want to get all crazy we could consider putting all the data into a [datascript](https://github.com/tonsky/datascript) and make that available to user hooks (to derive new data from existing ones) and to the pages, perhaps...

File: src/cryogen_core/compiler.clj [hunk 1]
```diff
@@ -468,8 +468,11 @@
    (when-not (empty? overrides)
      (println (yellow "overriding config.edn with:"))
      (pprint overrides))
-   (let [{:keys [^String site-url blog-prefix rss-name recent-posts keep-files ignored-files previews? author-root-uri theme]
-          :as   config} (resolve-config overrides)
+   (let [extend-params (or (:extend-params-fn overrides)
+                           (fn [params _] params))
+         overrides' (dissoc overrides)
```
Comment (lacarmen) [inline src/cryogen_core/compiler.clj]: missing a key here?

**Comment by lacarmen** [comment]
> Instead of adding :extend-params-fn to the overrides, we could pass it as an extra param 

Sorry, pass it in where? 

Cryogen has been pretty stable for a couple years now. Most issues I get are requests for some sort of custom functionality that relies on more data being available. So in general, I like the idea and definitely prefer it over using something heavier like datascript but perhaps it's something we could revisit. Happy to discuss ideas with you. :)

File: src/cryogen_core/compiler.clj (hunk 1 above)
Comment (holyjak) [reply src/cryogen_core/compiler.clj]: OMG! Fixed.

**Comment by holyjak** [comment]
> > Instead of adding :extend-params-fn to the overrides, we could pass it as an extra param
> 
> Sorry, pass it in where?

As a 2nd param to `compile-assets` and `compile-assets-timed` though if you are OK with adding a key like this, I think that is optimal for now,
>

> So in general, I like the idea and definitely prefer it over using something heavier like datascript but perhaps it's something we could revisit. Happy to discuss ideas with you. :)

I agree! If I ever get to it I can try to make a proof of concept with DataScript to see whether it is worth it. But for now, I prefer the smallest working change.

**Comment by lacarmen** [comment]
Sounds good :)

**Comment by holyjak** [comment]
@lacarmen What do you think needs to be done so that we can merge this PR?

**Comment by lacarmen** [comment]
@holyjak I guess I'm okay with passing the extra param :)

**Comment by lacarmen** [comment]
Currently implementation is fine too, I don't see a big need to change that for now :) I don't think there's anything else to be done then.

**Pull request merged**